File: DesignPatterns/01_GOF23/DP_examples/Python/02_Structural/Composite/Component.py
from abc import ABC, abstractmethod

# 叶子接口
class ILeaf(ABC):
    @abstractmethod
    def is_leaf(self):
        pass
    
    @abstractmethod
    def operation(self):
        pass

# 组件接口
class IComponent(ILeaf):
    @abstractmethod
    def add(self, component):
        pass
    
    @abstractmethod
    def remove(self, component):
        pass
    
    @abstractmethod
    def get_child(self, index):
        pass

# 组合类
class Composite(IComponent):
    def __init__(self, is_leaf):
        self._is_leaf = is_leaf
        self._sub_composites = []
    
    def is_leaf(self):
        return self._is_leaf
    
    def operation(self):
        if not self._is_leaf:
            for comp in self._sub_composites:
                comp.operation()
    
    def add(self, component):
        if self._is_leaf or component is None:
            return -1
        # 检查是否已存在
        if component in self._sub_composites:
            return -1
        self._sub_composites.append(component)
        return len(self._sub_composites)
    
    def remove(self, component):
        if self._is_leaf:
            return False
        if component in self._sub_composites:
            self._sub_composites.remove(component)
            return True
        return False
    
    def get_child(self, index):
        if index < 0 or index >= len(self._sub_composites):
            return None
        return self._sub_composites[index]

     

File: DesignPatterns/01_GOF23/DP_examples/Rust/02_Structural/Composite/component.rs
// 叶子接口
pub trait ILeaf {
    fn is_leaf(&self) -> bool;
    fn operation(&self);
}

// 组件接口
pub trait IComponent: ILeaf {
    fn add(&mut self, component: Box<dyn IComponent>) -> i32;
    fn remove(&mut self, component: &Box<dyn IComponent>) -> bool;
    fn get_child(&self, index: usize) -> Option<&Box<dyn IComponent>>;
}

// 组合类
pub struct Composite {
    is_leaf: bool,
    pub sub_composites: Vec<Box<dyn IComponent>>,
}

impl Composite {
    pub fn new(is_leaf: bool) -> Self {
        Self {
            is_leaf,
            sub_composites: Vec::new(),
        }
    }
}

impl ILeaf for Composite {
    fn is_leaf(&self) -> bool {
        self.is_leaf
    }

    fn operation(&self) {
        if !self.is_leaf {
            for comp in &self.sub_composites {
                comp.operation();
            }
        }
    }
}

impl IComponent for Composite {
    fn add(&mut self, component: Box<dyn IComponent>) -> i32 {
        if self.is_leaf {
            return -1;
        }

        // 检查是否已存在
        for comp in &self.sub_composites {
            // 这里简化处理，实际应该实现 PartialEq trait
        }

        self.sub_composites.push(component);
        self.sub_composites.len() as i32
    }

    fn remove(&mut self, component: &Box<dyn IComponent>) -> bool {
        if self.is_leaf {
            return false;
        }

        if let Some(index) = self.sub_composites.iter().position(|comp| {
            false
        }) {
            self.sub_composites.remove(index);
            return true;
        }
        false
    }

    fn get_child(&self, index: usize) -> Option<&Box<dyn IComponent>> {
        if index >= self.sub_composites.len() {
            None
        } else {
            Some(&self.sub_composites[index])
        }
    }
}
